### ISSUE TITLE
Bam.Core.Settings.InitializeAllInterfaces is slow on OSX Mono

### ISSUE THREAD

**Issue by markfinal** [markfinal/BuildAMation#29]
On large projects (many modules), this function is slow on Mono. It shows up with --stats on the PopulateGraph entry.

Some numeric examples from a project - on a Windows 8 VM, this can take < 1 second, while on the native host Mac, it can take 8 seconds.

Have tried replacing the MethodInfo Invoke() call with bound delegates but this made no difference, although there are some references (http://codeblog.jonskeet.uk/2008/08/09/making-reflection-fly-and-exploring-delegates/) that suggest this should have made a difference.

**Comment by markfinal** [comment]
This isn't showing up on Ubuntu VMs, both with a Windows 8 host, and a Yosemite host. So seems to be OSX specific.

**Comment by markfinal** [comment]
I'm seeing different results now, with the bottleneck elsewhere.

This is a Mac (Xcode) build of a reasonably large project (Player)

```
Task timing
-----------------------------------------------------------------------------------
Profile                  | Minutes | Seconds | Milliseconds |     Start |      Stop
-----------------------------------------------------------------------------------
ProcessCommandLine       |       0 |       0 |           65 | 34:41.873 | 34:41.939
GatherSource             |       0 |       6 |          264 | 34:41.946 | 34:48.210
AssemblyCompilation      |       0 |       0 |          806 | 34:48.210 | 34:49.017
LoadAssembly             |       0 |       0 |            1 | 34:49.017 | 34:49.018
PackageMetaData          |       0 |       0 |          328 | 34:49.018 | 34:49.347
IdentifyBuildableModules |       0 |       8 |          904 | 34:49.347 | 34:58.251
PopulateGraph            |       0 |       1 |          336 | 34:58.251 | 34:59.588
ValidateGraph            |       0 |       0 |            8 | 34:59.588 | 34:59.596
CreatePatches            |       0 |       2 |          130 | 34:59.596 |  35:1.726
ParseTokenizedStrings    |       0 |       0 |           89 |  35:1.726 |  35:1.816
GraphExecution           |       0 |       2 |          387 |  35:1.816 |  35:4.203
-----------------------------------------------------------------------------------
CumulativeTotal          |       0 |      22 |          321
-----------------------------------------------------------------------------------
TimedTotal               |       0 |      22 |          337 | 34:41.866 |  35:4.203
-----------------------------------------------------------------------------------
```

This is the Windows (VSSolution) (VM) running on the same hardware

```
Task timing
-----------------------------------------------------------------------------------
Profile                  | Minutes | Seconds | Milliseconds |     Start |      Stop
-----------------------------------------------------------------------------------
ProcessCommandLine       |       0 |       0 |           46 | 35:12.491 | 35:12.537
GatherSource             |       0 |       6 |          968 | 35:12.553 | 35:19.521
AssemblyCompilation      |       0 |       1 |           31 | 35:19.521 | 35:20.552
LoadAssembly             |       0 |       0 |           31 | 35:20.568 | 35:20.600
PackageMetaData          |       0 |       0 |           47 | 35:20.600 | 35:20.647
IdentifyBuildableModules |       0 |       1 |          609 | 35:20.647 | 35:22.257
PopulateGraph            |       0 |       1 |          280 | 35:22.257 | 35:23.537
ValidateGraph            |       0 |       0 |           15 | 35:23.537 | 35:23.552
CreatePatches            |       0 |       3 |          578 | 35:23.552 | 35:27.131
ParseTokenizedStrings    |       0 |       0 |           62 | 35:27.131 | 35:27.194
GraphExecution           |       0 |       1 |          624 | 35:27.194 | 35:28.818
-----------------------------------------------------------------------------------
CumulativeTotal          |       0 |      16 |          295
-----------------------------------------------------------------------------------
TimedTotal               |       0 |      16 |          326 | 35:12.491 | 35:28.818
-----------------------------------------------------------------------------------
```

**Comment by markfinal** [comment]
Further investigation is showing that the 

var module = new T();

call in Module.Create<T>() is very very slow on OSX, even on the simplest of Module types.

This is showing up on the simplest of examples (e.g. Test3).

**Comment by markfinal** [comment]
Further note that if I change Module.Create<T>() to take two template arguments, e.g. Module.Create<T,S>(), of which S just requires to be new'd, then if that argument is encoded to be Bam.Core.StringArray, and Module.Create calls

var array = new S();

(it also stores the array so it is not eliminated), then there is virtually no cost for the creation of S.

This rules out generics as the meaning.
It also rules out that the Module.Create call is created as a function through reflection (Graph.MakeModuleOfType).

To me, it suggests that creating a type through an assembly that is loaded explicitly, rather than as a reference, is very slow.

**Comment by markfinal** [comment]
This might be something to do with loading the package assembly into the bam command line assembly.

I suspect there is something going on with 'hidden' AppDomains when I do that, as there is apparently a significant cost for marshalling objects between domains.

A very simple test case, in which I load a prebuilt assembly and instantiate an object from that assembly, with this code

```
            var domain = System.AppDomain.CreateDomain("Remote");
            var assembly = domain.Load(asmBytes);

```
or

`            System.Reflection.Assembly assembly = System.Reflection.Assembly.Load(asmBytes);
`

demonstrated that the former has costs of around 4500 ticks, while the latter is around 33,000 ticks.

Changing bam to do this isn't trivial.

I suspected that the debug package assembly that I create would fair better, as it's a pre-packaged assembly without any assembly loads. However, with the timing I have on Module.Create,  it is no better. That was very unexpected.

**Comment by markfinal** [comment]
This may be to do with capturing stack traces in TokenizedString.

**Comment by markfinal** [comment]
I believe this is now fixed in #373 by disabling StackTrace capturing by default
